src/map.ts

```diff
@@ -10,34 +10,70 @@ export interface MapRect {
 export class MapLocation {
     constructor(public column: number, public row: number) {
     }
-}
 
-export class MapObject extends MapLocation {
+    setLocation(column: number, row: number) {
+        this.column = column;
+        this.row = row;
+    }
 }
 
-export class MapQuadrant {
-    protected objects: MapObject[];
-    protected data: number[][];
+export class MapObject extends MapLocation implements MapRect {
+    protected static nextID = 0;
+    protected static getID() {
+        return MapObject.nextID ++;
+    }
 
-    constructor() {
-        this.data = [];
-        this.objects = [];
+    public readonly id: number;
+    public width = 1;
+    public height = 1;
+
+    constructor(column = 0, row = 0) {
+        super(column, row);
+
+        this.id = MapObject.getID();
     }
 
-    addObject(obj: MapObject) {
-        this.objects.push(obj);
+    get left() {
+        return this.column;
     }
 
-    removeObject(obj: MapObject) {
-        const i = this.objects.indexOf(obj);
+    get top() {
+        return this.row;
+    }
 
-        if (i >= 0) {
-            this.objects.splice(i, 1);
-        }
+    get right() {
+        return this.column + this.width - 1;
+    }
+
+    get bottom() {
+        return this.row + this.height - 1;
+    }
+}
+
+export class MapArea extends MapObject {
+    primaryColor = "#00a1f2";
+    outlineColor = "#00a1f2";
+
+    constructor(public width = 1, public height = 1) {
+        super(0, 0);
     }
 
-    getObjects() {
-        return this.objects.slice();
+    setColors(primary: string, outline: string) {
+        this.primaryColor = primary;
+        this.outlineColor = outline;
+    }
+
+    setDimensions(width: number, height: number) {
+        if (width > 0) this.width = width | 0;
```

Comment:
Completely harmless, but if i remember right any comparison with undefined or null results in false so the `| 0`s here should never do anything?

Reply:
It converts from a float to an integer